### ISSUE TITLE
Crash when sourcing from large JSON file

### ISSUE THREAD

**Issue by willmartian** [11ty/eleventy#1597]
**Describe the bug**
I am trying to paginate over a pretty large JSON file, 17.2MB.  Eleventy crashes when ran. Any advice?

**Expected behavior**
Output of paginated HTML files in the build folder. 

**Error Message**
```
<--- Last few GCs --->

[22797:0x1f43a50]     6587 ms: Mark-sweep 1259.2 (1457.7) -> 1258.7 (1457.7) MB, 105.9 / 0.0 ms  (average mu = 0.202, current mu = 0.049) allocation failure scavenge might not succeed
[22797:0x1f43a50]     6989 ms: Mark-sweep 1259.3 (1457.7) -> 1259.2 (1457.7) MB, 399.9 / 0.0 ms  (average mu = 0.067, current mu = 0.007) allocation failure scavenge might not succeed


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0xc6ef995452b]
    1: StubFrame [pc: 0xc6ef99556f3]
Security context: 0x02ac1e2aee11 <JSObject>
    2: normalize [0x27b150b9a359] [path.js:~1111] [pc=0xc6ef9892c5f](this=0x34884c282389 <Object map = 0x3b1419dcab89>,path=0x02ec0cb71549 <String[32]: /home/will/Projects/myRedactedProjectName>)
    3: getLocalData [0x19ccc3cfda89] [/usr/local/lib/node_modules/@11ty/eleventy/src/TemplateData.js:~280] [pc=0xc6ef9a25130](this=0x2d6b...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x7f74b00dc46c node::Abort() [/lib/x86_64-linux-gnu/libnode.so.64]
 2: 0x7f74b00dc4b5  [/lib/x86_64-linux-gnu/libnode.so.64]
 3: 0x7f74b0308e6a v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/lib/x86_64-linux-gnu/libnode.so.64]
 4: 0x7f74b03090e1 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/lib/x86_64-linux-gnu/libnode.so.64]
 5: 0x7f74b06a3c66  [/lib/x86_64-linux-gnu/libnode.so.64]
 6: 0x7f74b06b5043 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/lib/x86_64-linux-gnu/libnode.so.64]
 7: 0x7f74b06b5930 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/lib/x86_64-linux-gnu/libnode.so.64]
 8: 0x7f74b06b791d v8::internal::Heap::AllocateRawWithLigthRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/lib/x86_64-linux-gnu/libnode.so.64]
 9: 0x7f74b06b7975 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/lib/x86_64-linux-gnu/libnode.so.64]
10: 0x7f74b0683dda v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/lib/x86_64-linux-gnu/libnode.so.64]
11: 0x7f74b090f31e v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/lib/x86_64-linux-gnu/libnode.so.64]
12: 0xc6ef995452b 
Aborted (core dumped)
```
**Environment:**

- OS and Version: Ubuntu Linux 20.04.1 LTS, 32GB RAM
- Eleventy Version: 0.11.1

**Comment by willmartian** [comment]
I think this is a similar issue with #695 . However, increasing the --max-old-space-size flag did not fix the error for me. I am using Node v15.6.0.

**Comment by pdehaan** [comment]
Interesting. I was able to parse a 34.99 MB JSON file in my site's local /src/_data/ folder [on a ~2 yr old macOS laptop] w/ 11ty v0.11.1; and it was able to generate a 17825 line, 815 KB /index.html file in around 1.42 seconds.

Although it's literally a very oversimplified test case w/ 1 data file and 1 LiquidJS template. No fancy nested content or includes or partials or layouts.

```sh
 npm run build

> 11ty-big-data@1.0.0 build /Volumes/Dev/github/pdehaan/11ty-big-data
> eleventy

src/_data/untriaged.json is 34.99 MB
Writing www/index.html from ./src/pages/home.liquid.
Wrote 1 file in 1.42 seconds (v0.11.1)
```

Admittedly, not the most helpful response you were hoping for, but curious if that means it's possibly a Linux issue, although I think a few people cited macOS in #695 as well, so possibly not.

I could try a similar test on a 16GB M1 Mac mini and see if maybe it's more of a memory constraint (although the mini is much newer hardware so it might handle memory differently with the new fancy chip).


**Environment:**

- OS and Version: macOS 10.15.7 (32 GB RAM)
- Eleventy Version: 0.11.1

**Comment by pdehaan** [comment]
Here's the same test on the newer Mac mini w/ M1 and 16 GB RAM. Interestingly it was &frac12; the time to build the site/page (although not too surprising considering there isn't anything else on the computer except Node v15.5.1 and VS Code).

```sh
npm run build

> 11ty-big-json@1.0.0 build
> eleventy

src/_data/untriaged.json is 34.99 MB
Writing www/index.html from ./src/pages/home.liquid.
Wrote 1 file in 0.71 seconds (v0.11.1)
```

---

Just for fun, I found a bigger JSON file, which weighs in at 138 MB, and nearly 5 million lines of JSON junk. It still managed to build the site in _just_ under 3 seconds. 

```sh
wc -l _data/components/general.json # 4,947,832
npm run build

> 11ty-big-json@1.0.0 build
> eleventy

src/_data/general.json is 138.24 MB
Writing www/index.html from ./src/pages/home.liquid.
Wrote 1 file in 2.99 seconds (v0.11.1)
```

**Environment:**

- OS and Version: macOS 11.1 (Big Sur; 16 GB RAM)
- Eleventy Version: 0.11.1

**Comment by willmartian** [comment]
Thanks for looking into this! My machine is 32 GB RAM, so maybe that is pointing towards a Linux issue? I will try running the project on Windows and report back.

**Comment by Ryuno-Ki** [comment]
I'm running a GNU/Linux machine (with 3.5 GB RAM). So it sounds like a good candidate :)
Does anybody have a link to a large JSON file?

(An option would be to stream the JSON instead of loading it all into memory - likely requiring some architectural changes on Eleventy's end)

**Comment by willmartian** [comment]
Okay, no luck on my Windows machine or on Netlify.

Here is the project I am working on, if anyone wants to try pulling it down. @Ryuno-Ki @pdehaan 
https://github.com/willmartian/trump-twitter

**Comment by pdehaan** [comment]
OK, well, _that_ crashes my Node. But fun to watch Activity Monitor and my Node process' memory usage jump from around 20MB up to 4.05GB before dying.

I've run this about a dozen times locally (Node 14 and Node 15) and the errors aren't always the same, but the general gist is:

```sh
FATAL ERROR: MarkCompactCollector: young object promotion failed Allocation failed - JavaScript heap out of memory
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
FATAL ERROR: MarkCompactCollector: young object promotion failed Allocation failed - JavaScript heap out of memory
FATAL ERROR: MarkCompactCollector: young object promotion failed Allocation failed - JavaScript heap out of memory
```

I'll see if the Mac mini is any luckier.

Next, I guess we'll see if I can do a vanilla Node test script which does a `require()` on the 16M /src/_data/tweets.json and count rows or something trivial. Otherwise, I guess we start cutting rows from the JSON file until it loads and see if we can find a threshold for where Node/Eleventy starts working again.

**Comment by pdehaan** [comment]
After re-reading the other bug, and trying some things, it's still crashing on my laptop, but it's using about 12 GB of memory now. I'd be reluctant to set that much higher and have it use 100% of my laptop's RAM.

```sh
node --max-old-space-size=12000 ./node_modules/@11ty/eleventy/cmd.js --quiet
```

---

```js
// test.js
const tweets = require("./src/_data/tweets.json");

console.log(tweets.length); // 56571
```

That runs near instantly. So I'm not sure it's an issue w/ Node opening/processing large data files. It might be more about 11ty's handling of extreme global data or trying to paginate over ~57k records.
Not sure how to mitigate that though, unless you try splitting the tweets.json into multiple global data files. But I can't guarantee that it'd work. It's still the same amount of data, but maybe Node's GC might handle it a bit better if it only has to paginate over 10k chunks. Although that makes your front end code a lot bit trickier.

---

OK, minor "progress".

```sh
node test # 18667 records


ls -lash src/_data 
   72 -rw-r--r--  1 pdehaan  wheel    33K 21 Jan 07:43 short.json
12544 -rw-r--r--  1 pdehaan  wheel   5.4M 21 Jan 08:21 tweets.json


npm run build

> trumptwitter@1.0.0 build /private/tmp/trump-twitter
> node --max-old-space-size=12000 ./node_modules/@11ty/eleventy/cmd.js --quiet

Problem writing Eleventy templates: (more in DEBUG output)
> Output conflict: multiple input files are writing to `_site/realDonaldTrump/status/1291488374835097600/data.json`. Use distinct `permalink` values to resolve this conflict.
  1. ./src/tweetData.11ty.js
  2. ./src/tweetData.11ty.js

`DuplicatePermalinkOutputError` was thrown:
    (Repeated output has been truncated…)
        at TemplateMap.checkForDuplicatePermalinks (/private/tmp/trump-twitter/node_modules/@11ty/eleventy/src/TemplateMap.js:547:13)
        at TemplateMap.cache (/private/tmp/trump-twitter/node_modules/@11ty/eleventy/src/TemplateMap.js:308:10)
        at async TemplateWriter._createTemplateMap (/private/tmp/trump-twitter/node_modules/@11ty/eleventy/src/TemplateWriter.js:170:5)
        at async TemplateWriter.writeTemplates (/private/tmp/trump-twitter/node_modules/@11ty/eleventy/src/TemplateWriter.js:203:5)
        at async TemplateWriter.write (/private/tmp/trump-twitter/node_modules/@11ty/eleventy/src/TemplateWriter.js:254:25)
        at async Eleventy.write (/private/tmp/trump-twitter/node_modules/@11ty/eleventy/src/Eleventy.js:743:13)
Wrote 0 files in 98.03 seconds (v0.11.1)
```

So, I removed around 38K tweets from tweets.json (and it's _still_ 5.4 MB) and increased my max memory allocation to 12 GB and it now starts to build, but ends up choking on a permalink conflict with the following file:

> _site/realDonaldTrump/status/1291488374835097600/data.json

**Comment by pdehaan** [comment]
Success.

```sh
npm run build     

> trumptwitter@1.0.0 build /private/tmp/trump-twitter
> node --max-old-space-size=12000 ./node_modules/@11ty/eleventy/cmd.js --quiet

Wrote 56000 files in 113.31 seconds (2.0ms each, v0.11.1)
```

---

Re: Twitter IDs,
https://developer.twitter.com/en/docs/twitter-ids was pretty interesting.

I think that's what's happening. All the tweet `id` properties are suspiciously round, so it's probably losing accuracy due to converting 64-bit integers to 53-bit integers.

```js
{"id":1262933597536899000,"text":...},
{"id":1262933225061695500,"text":...},
{"id":1262933129980973000,"text":...},
{"id":1262933117851091000,"text":...},
{"id":1262933104299212800,"text":...},
{"id":1262932924615319600,"text":...},
{"id":1223832546406011000,"text":...},
etc
```

They cite an `id` and `id_str` property from the Twitter v1.1 API, but I don't see that in the tweet.json file.

```json
{"id": 10765432100123456789, "id_str": "10765432100123456789"}
```

---

TLDR: I'm lost. It's a huge amount of data, and I'm not sure if it's just an "11ty-thing" where it's trying to add all the 16MB of tweets in each page's scope, or the fact it might be creating 57k records into an "all" collection (or maybe it's 171k items in a collection since there are pages for 'tweet's, 'tweetData', and 'tweetEmbed's).

I guess now that we've proven that it "works"(tm) via some magic of removing 67% of the records, the one [so far] duplicated tweet `id`, you could slowly add back more records until it starts failing. But because it's an out-of-memory error, it might always fail in slightly different places/records depending on the current state of your laptop's memory.

Only other idea I had was possibly seeing if it would be possible to do this with vanilla LiquidJS and not use 11ty. Basically just have LiquidJS templates for each page template (tweet, tweetData, tweetEmbed) and then do a single loop over the tweets.json and write the template output for each template type. Not sure how difficult it'd be to set up [LiquidJS](https://liquidjs.com/) partials and includes, but they have some pretty good docs. It's the same package that 11ty uses internally for .liquid template processing. But https://liquidjs.com/tutorials/render-file.html is probably a good start.

**Comment by Ryuno-Ki** [comment]
Tweets, huh?

I'm wondering, whether it might make sense to load them into a SQLite3 database first and then process them in chunks.
Could be friendlier to the memory then file I/O …

@willmartian What template language do you use? From the linked repo I'd guess template literals

**Comment by willmartian** [comment]
@pdehaan Many many thanks for investigating! I am still new to 11ty, and I am figuring out a lot of this as I go along. 

I think I have a solution? I used Eleventy inside of Eleventy. First, I split tweets.json into smaller JSONs, using a template with pagination:

![image](https://user-images.githubusercontent.com/17113462/105406111-faf0c700-5bf9-11eb-9766-93f969a5d3b5.png)

Then, I also generated multiple .11ty.js templates to correspond with the split data files:

![image](https://user-images.githubusercontent.com/17113462/105406228-1fe53a00-5bfa-11eb-822b-0b64b1cb5c8f.png)

And after running all of those generated files in my top-level Eleventy project... everything worked!

![image](https://user-images.githubusercontent.com/17113462/105408201-a438bc80-5bfc-11eb-80cc-dc1baa1b717b.png)

Granted, this seems a little hacky--but maybe it could be wrapped in a plugin or something as a workaround for generating lots and lots of pages.

<hr>

Also, great catch on the Twitter IDs. I will see if I can find a dataset that has the string equivalents.

**Comment by willmartian** [comment]
> Tweets, huh?
> 
> I'm wondering, whether it might make sense to load them into a SQLite3 database first and then process them in chunks.
> Could be friendlier to the memory then file I/O …
> 
> @willmartian What template language do you use? From the linked repo I'd guess template literals

Yeah, I think that is definitely a next step. To be fair, this whole project might also be better suited for a framework like Preact rather than trying to generate so many static files. But hey, I like Eleventy. :)

I use the JavaScript template literals! Mainly because I am familiar with JavaScript already. I might check out nunjucks as well, though, to learn something new.

**Comment by pdehaan** [comment]
> I am figuring out a lot of this as I go along.

I'm still figuring this out as I go along...

---


> First, I split tweets.json into smaller JSONs, using a template with pagination:

Great workaround!
I'll have to check it out when you push the changes to GitHub. I wouldn't have thought of using pagination that way. ❤️ 

Even more amazed at generating 170k files in around 3 minutes (only 1.1ms per file). 🏁 🔥 :fire:

**Comment by Ryuno-Ki** [comment]
> Yeah, I think that is definitely a next step. To be fair, this whole project might also be better suited for a framework like Preact rather than trying to generate so many static files. But hey, I like Eleventy. :)

Oh, I am doing similar stuff with my Twitter takeout (European citizen here. Do you have the option to download everything Twitter knows about you in other places of the world?).
That scratches more on that 36k limit we've seen above.

I can feel your pain :)

**Comment by pdehaan** [comment]
I was bored, here's a rough version of the site in LiquidJS: https://github.com/pdehaan/liquidjs-tweets

Basically uses vanilla Node+LiquidJS to loop over the tweets.json file and uses templates+blocks to write out the static site.

```sh
time npm run build

> liquidjs-tweets@1.0.0 build /Volumes/Dev/github/pdehaan/liquidjs-tweets
> node index

npm run build  59.12s user 57.60s system 104% cpu 1:51.60 total

tree www | grep -E "\.(html|json)$" | wc -l # 169,712
```

So, it can write ~170K files in under 2 minutes. Although it will take a lot longer if you run prettier against each file while writing it out (although prettier did find a few HTML errors with the current templates). I didn't check to see if it'd be any faster to run Prettier via the CLI as a "postbuild" hook.

But Liquid+Node had no problems processing the entire 16MB JSON file without needing to use streams or pagination or chunking. Plus, super easy to add a cheap `.slice(0, 50)` call in there if you only want to process 50 tweets as a dry-run instead of rebuilding the entire site each time (which is too slow to be fun).

**Comment by willmartian** [comment]
@pdehaan Just looked at your liquid tweets repo--it is so readable, simple, and clear what is going on! While I was able to get my nested workaround _somewhat_ organized (https://github.com/willmartian/trump-twitter), your solution seems better suited for my project at this time. 

If I am planning to continue developing from the code you have, would you prefer that I fork the repo to maintain the link to your contribution, or clone it and mention you in readme? Btw, I will be hosting the project at https://trumptwitter.org

**Comment by pdehaan** [comment]
@willmartian I like https://github.com/pdehaan/gray-matter-liquidjs-test. I rewrote the liquidjs-tweets repo from scratch last night and pushed it this morning and I think that's a much cleaner approach with support for some basic template front-matter and permalinks.

Re: developing,
Don't care. Copy all the code and release it under your name. I'm probably done with both repos and they'll just sit there and rot. No attribution needed.
If your GitHub repo is current, I can submit a few PRs to the repo for some HTML issues I found on the original templates, or I can wait for you to update the site if you're thinking of migrating to a different solution. Nothing big, just a few unclosed `</div>` tags, or Prettier complaining about closing `</input>` tags.

**Comment by zachleat** [comment]
Howdy y’all—I think there are a couple of things that are probably worth noting here that improved with the 1.0 release—although I’m not sure if this project is still active? I didn’t see the downstream github repo at all.

1. Pagination didn’t used to save a cached version of the compiled template. It does now.
2. Template and directory data files used to run N times per paginated output. We cache those now.

I think 1.0 stable would probably handle this better—otherwise if this project is now archived maybe we can close this one

**Comment by willmartian** [comment]
Hi @zachleat 😄

This is no longer an active project, and I think the issue is a fairly niche one. Given the number of pages being generated, I think if I were to move forward with a large archival project like this in the future I would try out 11ty Serverless with Netlify's on-demand builders. With that in mind I am going to go ahead and close.

(Also, wow, this issue being a year old reminded me I have been using and having fun with 11ty for a while now! Thanks for all the hard work and congrats on 1.0).
